vault/identity_store_oidc_provider.go

```diff
@@ -345,3 +520,174 @@ func (i *IdentityStore) pathOIDCScopeExistenceCheck(ctx context.Context, req *lo
+		if err != nil {
+			return nil, err
+		}
+		if entry == nil {
+			return logical.ErrorResponse("cannot find assignment %q", assignment), nil
+		}
+	}
+
+	if keyRaw, ok := d.GetOk("key"); ok {
+		key := keyRaw.(string)
+		if req.Operation == logical.UpdateOperation && client.Key != key {
+			return logical.ErrorResponse("key modification is not allowed"), nil
+		}
+		client.Key = keyRaw.(string)
+	} else if req.Operation == logical.CreateOperation {
+		client.Key = d.Get("key").(string)
+	}
+
+	if client.Key == "" {
+		return logical.ErrorResponse("the key parameter is required"), nil
+	}
+
+	// enforce key existence on client creation
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+client.Key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("cannot find key %q", client.Key), nil
+	}
+
+	if idTokenTTLRaw, ok := d.GetOk("id_token_ttl"); ok {
+		client.IDTokenTTL = idTokenTTLRaw.(int)
+	} else if req.Operation == logical.CreateOperation {
+		client.IDTokenTTL = d.Get("id_token_ttl").(int)
+	}
+
+	if accessTokenTTLRaw, ok := d.GetOk("access_token_ttl"); ok {
+		client.AccessTokenTTL = accessTokenTTLRaw.(int)
+	} else if req.Operation == logical.CreateOperation {
+		client.AccessTokenTTL = d.Get("access_token_ttl").(int)
+	}
+
+	if clientID, ok := d.GetOk("client_id"); ok {
+		client.ClientID = clientID.(string)
+	}
+
+	if client.ClientID == "" {
+		// generate client_id
+		clientID, err := base62.Random(32)
+		if err != nil {
+			return nil, err
+		}
+		client.ClientID = clientID
+	}
+
+	if clientSecret, ok := d.GetOk("client_secret"); ok {
+		client.ClientSecret = clientSecret.(string)
+	}
+
+	if client.ClientSecret == "" {
+		// generate client_secret
+		clientSecret, err := base62.Random(64)
+		if err != nil {
+			return nil, err
+		}
+		client.ClientSecret = clientSecret
+	}
+
+	// store client
+	entry, err = logical.StorageEntryJSON(clientPath+name, client)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// pathOIDCListClient is used to list clients
+func (i *IdentityStore) pathOIDCListClient(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	clients, err := req.Storage.List(ctx, clientPath)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(clients), nil
+}
+
+// pathOIDCReadClient is used to read an existing client
+func (i *IdentityStore) pathOIDCReadClient(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, clientPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var client client
+	if err := entry.DecodeJSON(&client); err != nil {
+		return nil, err
+	}
+	return &logical.Response{
+		Data: map[string]interface{}{
```

Comment:
Q: should we be returning the `client_id` and `client_secret` for read operations?

Reply:
cc @austingebauer

Reply:
Yep, we should return both the `client_id` and `client_secret` so that they can be used to configured OIDC clients.